modules/diskq/qdisk.c

```diff
@@ -203,6 +205,14 @@ qdisk_is_space_avail(QDisk *self, gint at_least)
 static void
 _truncate_file(QDisk *self, off_t expected_size)
 {
+  off_t truncate_threshold = (off_t)(qdisk_get_maximum_size(self) * TRUNCATE_SIZE_RATIO);
+  off_t possible_size_reduction = self->file_size - expected_size;
+
+  if (possible_size_reduction >= 0 && possible_size_reduction < truncate_threshold)
```

Comment:
Actually that condition provided that we call `ftruncate()` if we want to extend the file with it.

Though, I agree that it is hard to think through in the current form, so I made small helper functions for the conditions. I hope it will make it more readable.